File: python/nav_study/global_planning/astar.py
"""A* search — f = g + w*h with an admissible grid heuristic.

Hart, Nilsson & Raphael (1968). w = 1 is optimal; w > 1 is weighted A* (Pohl
1970): faster, bounded-suboptimal. w is read from the `heuristic_weight` param.
"""

from __future__ import annotations

from nav_study.core.params import ParamSet

from ._bestfirst import _BestFirstSearch


class AStar(_BestFirstSearch):
    def __init__(self, params: ParamSet) -> None:
        super().__init__(params)
        self._heuristic_weight = params.get_float("heuristic_weight")

    @property
    def name(self) -> str:
        return "astar"

    def _uses_heuristic(self) -> bool:
        return True

    def _weight(self) -> float:
        return self._heuristic_weight

File: cpp/src/global_planning/astar.cpp
#include "nav_study/global_planning/astar.hpp"

#include "nav_study/global_planning/discrete_search.hpp"

namespace nav_study::global_planning {

core::PlanResult<Cell> AstarPlanner::plan(DiscreteSpace<Cell>& space, const Cell& start,
                                          const Cell& goal, TraceRecorder* recorder) {
  // Weighted A*: f = g + w*h. w == 1 stays admissible/optimal (Pohl 1970).
  double w = params_.get_float("heuristic_weight");
  return best_first_search(space, start, goal, recorder,
                           [&](const Cell& c) { return w * space.heuristic(c, goal); });
}

}  // namespace nav_study::global_planning
